pkg/ipcache/metadata.go

```diff
@@ -148,6 +148,8 @@ func InjectLabels(src source.Source, updater identityUpdater, triggerer policyTr
 			}
 		}
 	}
+	// Don't hold lock while calling UpdateIdentities, as it will otherwise run into a deadlock
```

Comment:
(unrelated to this line)

I'm curious how you were able to reproduce this and whether there's an opportunity to introduce any tests to cover this? Clearly we didn't catch this the first time, and given that this is a lock coordination issue across multiple goroutines in separate subsystems, it can be a little tricky to identify purely through code review of specific areas.

Reply:
I encountered the deadlock in the same situation as described in #18352 (high load, many nodes starting up and getting scheduled with 100% IP use immediately due to many pending PODs). The symptom was that something (I don't remember what exactly) was suddenly stuck on many nodes and that listing endpoints via cilium cli was also hangling. I then used some go debugging tool (also forget the name of it...can't check atm) to dump all stack traces.  Based on the call stacks of I discovered the 2 offending locks, after which I started analysing the corresponding code.

I don't have a reliable reproducer unfortunately.